### PR TITLE
[feature/rust_vmm_adoption] Integrate seccompiler

### PR DESCRIPTION
**Pull request by alindima**
**This PR is currently made on a feature branch. We should only integrate it in `main` when approaching the v2.0.0 release because it removes the deprecated `--basic` seccompiler-bin parameter, which is a breaking change.**

# Reason for This PR

We should consume the rust-vmm code for seccompiler, since we've upstreamed it and released it there.

## Description of Changes

View commit messages.
Notably, we remove the build-time dependency on seccompiler-bin. We instead include seccompiler as a build script library dependency. It removes the overhead associated with deserialising the filter with bincode on startup. This also enables us to get rid of some not-so-pretty hacks in the tests and the build scripts.

Also note the coverage drop. It may look big, but it's due to removing a good chunk of well-covered code.

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: src/api_server/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ serde_json = ">=1.0.9"
 logger = { path = "../logger" }
 micro_http = { git = "https://github.com/firecracker-microvm/micro-http", rev = "0a58eb1" }
 mmds = { path = "../mmds" }
-seccompiler = { path = "../seccompiler" }
+seccompiler =  "^0.2.0"
```
Comment (AlexandruCihodaru) [inline src/api_server/Cargo.toml:17]: should we use instead >= 0.2.0?

File: src/api_server/src/lib.rs [hunk 2]
```diff
@@ -189,11 +187,13 @@ impl ApiServer {
         // Load seccomp filters on the API thread.
         // Execution panics if filters cannot be loaded, use --no-seccomp if skipping filters
         // altogether is the desired behaviour.
-        if let Err(e) = seccompiler::apply_filter(seccomp_filter) {
-            panic!(
-                "Failed to set the requested seccomp filters on the API thread: Error: {:?}",
-                e
-            );
+        if let Some(filter) = seccomp_filter {
+            if let Err(e) = seccompiler::apply_filter(&*filter) {
```
Comment (AlexandruCihodaru) [inline src/api_server/src/lib.rs:191]: no need to dereference it `&filter` should work

File: src/seccompiler-bin/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+                .help("File path of the JSON input."),
+        )
+        .arg(
+            Argument::new("output-file")
+                .required(false)
+                .takes_value(true)
+                .default_value(DEFAULT_OUTPUT_FILENAME)
+                .help("Optional path of the output file."),
+        )
+        .arg(
```
Comment (AlexandruCihodaru) [inline src/seccompiler-bin/src/main.rs:80]: just an idea but what if we would use a default value for this as well. DEFAULT_TARGET which is the current target and we define the variable using target_arch = "..."

File: src/utils/src/seccomp.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#[repr(C)]
+#[derive(Serialize, Deserialize)]
+/// We need to reimplement this type locally so that we may derive
+/// `Serialize` and `Deserialize` for it.
+pub struct sock_filter {
+    pub code: libc::c_ushort,
+    pub jt: libc::c_uchar,
+    pub jf: libc::c_uchar,
+    pub k: libc::c_uint,
+}
```
Comment (AlexandruCihodaru) [inline src/utils/src/seccomp.rs:23]: Can we use the full name for these?

**Review by AlexandruCihodaru — COMMENTED** [review]
Part 1. First pass through the PR, so far it look great :D

File: docs/seccompiler-bin.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    --target-arch "x86_64"  # The CPU arch where the BPF program will run.
+                            # Supported architectures: x86_64, aarch64.
+    --input-file "x86_64_musl.json" # File path of the JSON input.
+    --output-file "bpf_x86_64_musl" # Optional path of the output file.
+                                    # [default: "seccomp_binary_filter.out"]
+```
+
+## Where is seccompiler-bin implemented?
+
+Seccompiler is implemented as another package in the Firecracker cargo
```
Comment (dianpopa) [inline docs/seccompiler-bin.md:34]: secompiler-bin is implemented...

File: docs/seccompiler-bin.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+workspace. The code is located at `firecracker/src/seccompiler/src`.
```
Comment (dianpopa) [inline docs/seccompiler-bin.md:35]: `firecracker/src/seccompiler-bin/src`

File: src/seccompiler-bin/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,456 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (alindima) [reply src/seccompiler-bin/src/main.rs:1]: I think Github only sees renames when there are not other changes to the file.
In this case there are actually differences between the files

File: src/seccompiler-bin/src/main.rs (hunk 3 above)
Comment (alindima) [reply src/seccompiler-bin/src/main.rs:80]: I wouldn't do this. Users may then become unaware that these filters are arch-dependent and may overlook this, resulting in catastrophic runtime errors when executing the filter.

File: src/utils/Cargo.toml [hunk 8]
```diff
@@ -9,6 +9,8 @@ license = "Apache-2.0"
 libc = ">=0.2.39"
 serde = { version = ">=1.0.27", features = ["derive"] }
 vmm-sys-util = ">=0.8.0"
+seccompiler = "0.2.0"
```
[the remark this answers is not in the document]
Comment (alindima) [reply src/utils/Cargo.toml:12]: done

File: src/utils/Cargo.toml [hunk 9]
```diff
@@ -9,6 +9,8 @@ license = "Apache-2.0"
...
+bincode = "1.2.1"
```
[the remark this answers is not in the document]
Comment (alindima) [reply src/utils/Cargo.toml:13]: done

File: src/vmm/src/vstate/vcpu/mod.rs [hunk 10]
```diff
@@ -264,15 +263,17 @@ impl Vcpu {
...
         // altogether is the desired behaviour.
-        if let Err(e) = seccompiler::apply_filter(seccomp_filter) {
-            panic!(
-                "Failed to set the requested seccomp filters on vCPU {}: Error: {}",
-                self.kvm_vcpu.index, e
-            );
+        if let Some(filter) = seccomp_filter {
+            if let Err(e) = seccompiler::apply_filter(&*filter) {
+                panic!(
+                    "Failed to set the requested seccomp filters on vCPU {}: Error: {}",
```
[the remark this answers is not in the document]
Comment (alindima) [reply src/vmm/src/vstate/vcpu/mod.rs:273]: Done

File: docs/seccompiler-bin.md (hunk 5 above)
Comment (alindima) [reply docs/seccompiler-bin.md:34]: done

File: docs/seccompiler-bin.md (hunk 6 above)
Comment (alindima) [reply docs/seccompiler-bin.md:35]: done

**Comment by alindima** [comment]
@dianpopa @AlexandruCihodaru I've addressed all of your comments, PTAL.
I also added a new commit where I remove some hardcoded filters from the Firecracker unit tests and instead use the seccompiler lib for compiling them

**Review by dianpopa — APPROVED** [review]

**Pull request merged**